controllers/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +56,533 @@ type VolumeGroupReplicationReconciler struct {
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("volumeGroupReplication resource not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Get VolumeGroupReplicationClass instance
+	vgrClassObj, err := r.getVolumeGroupReplicationClass(logger, instance.Spec.VolumeGroupReplicationClassName)
+	if err != nil {
+		logger.Error(err, "failed to get volumeGroupReplicationClass resource", "VGRClassName", instance.Spec.VolumeGroupReplicationClassName)
+		err := r.setGroupReplicationFailure(instance, logger, err)
+		return reconcile.Result{}, err
+	}
+
+	err = validatePrefixedParameters(vgrClassObj.Spec.Parameters)
+	if err != nil {
+		logger.Error(err, "failed to validate parameters of volumeGroupReplicationClass", "VGRClassName", instance.Spec.VolumeGroupReplicationClassName)
+		err := r.setGroupReplicationFailure(instance, logger, err)
+		return reconcile.Result{}, err
+	}
+
+	// Declare all dependent resources
+	vgrContentObj := &replicationv1alpha1.VolumeGroupReplicationContent{}
+	vrObj := &replicationv1alpha1.VolumeReplication{}
+
+	// Create/Update dependent resources only if the instance is not marked for deletion
+	if instance.GetDeletionTimestamp().IsZero() {
+
+		// Add finalizer to VGR instance
+		if err = AddFinalizerToVGR(r.Client, logger, instance); err != nil {
+			logger.Error(err, "failed to add VolumeGroupReplication finalizer")
+			return reconcile.Result{}, err
+		}
+
+		// Check if PVCs exist based on provided selectors
+		pvcList := corev1.PersistentVolumeClaimList{}
+		if instance.Spec.Source.Selector != nil {
+			labelSelector := ""
+			pvcList, labelSelector, err = r.getMatchingPVCsFromSource(instance, logger)
+			if err != nil {
+				logger.Error(err, "failed to get PVCs using selector")
+				err := r.setGroupReplicationFailure(instance, logger, err)
+				return reconcile.Result{}, err
+			}
+			if len(pvcList.Items) == 0 {
+				logger.Info("No matching PVCs found for the given selectors")
+				return reconcile.Result{}, nil
+			}
+
+			// Add the string representation of the labelSelector to the VGR annotation
+			if instance.ObjectMeta.Annotations == nil {
+				instance.ObjectMeta.Annotations = make(map[string]string)
+			}
+
+			if instance.ObjectMeta.Annotations["pvcSelector"] != labelSelector {
+				instance.ObjectMeta.Annotations["pvcSelector"] = labelSelector
+				err = r.Client.Update(ctx, instance)
+				if err != nil {
+					logger.Error(err, "failed to add pvc selector annotation to VGR")
+					err := r.setGroupReplicationFailure(instance, logger, err)
+					return reconcile.Result{}, err
+				}
+			}
+		} else {
+			logger.Info("No selector provided for PVCs in volumeGroupReplication source")
+			return reconcile.Result{}, nil
+		}
+
+		// Update PersistentVolumeClaimsRefList in VGR Status
+		tmpRefList := []corev1.LocalObjectReference{}
+		for i, pvc := range pvcList.Items {
+			if i < 100 {
```

Comment:
add this check above if we have more than 100 PVC after getMatchingPVCsFromSource